### DISCUSSION TITLE
http status 404

### DISCUSSION THREAD

**Discussion by ameknite** [rust-lang/crates.io#8307] [General]
Can someone explain me why when I try to get the header of a crate page, it gives me `HTTP/2 404 `
Is this a bug?

example:
`curl -I https://crates.io/crates/rand` 
gives you:
```
HTTP/2 404 
content-length: 0
server: Cowboy
report-to: {"group":"heroku-nel","max_age":3600,"endpoints":[{"url":"https://nel.heroku.com/reports?ts=1710550532&sid=af571f24-03ee-46d1-9f90-ab9030c2c74c&s=494ZKz8htAtYLT%2Bl9FGg7rswx3Bpw7WmWdxkk0p8yJM%3D"}]}
reporting-endpoints: heroku-nel=https://nel.heroku.com/reports?ts=1710550532&sid=af571f24-03ee-46d1-9f90-ab9030c2c74c&s=494ZKz8htAtYLT%2Bl9FGg7rswx3Bpw7WmWdxkk0p8yJM%3D
nel: {"report_to":"heroku-nel","max_age":3600,"success_fraction":0.005,"failure_fraction":0.05,"response_headers":["Via"]}
access-control-allow-origin: *
strict-transport-security: max-age=31536000; includeSubDomains
date: Sat, 16 Mar 2024 00:55:32 GMT
via: 1.1 vegur, 1.1 f1c5a27e35ac834b32534347282b3496.cloudfront.net (CloudFront)
x-cache: Error from cloudfront
x-amz-cf-pop: DFW55-C1
x-amz-cf-id: Lr0sdRMwYLw_lRqhm7Wv9wZTv1FxaSplOKfa1zQaITFl6U3jZXHDrQ==
```

**Comment by LawnGnome** [comment]
> Is this a bug?

This is intentional.

The crates.io frontend is implemented using Ember.js, so there's no backend route that returns the HTML that you see for, say, https://crates.io/crates/rand. What happens instead is that requests with `Accept: text/html` (such as any request your browser would send) is intercepted within the crates.io backend by [this middleware](https://github.com/rust-lang/crates.io/blob/main/src/middleware/ember_html.rs), which then serves up the Ember.js bundle. Requests that don't have that `Accept` header (like this `curl` command) get a 404 instead, as you saw.

I'm guessing that what you're probably more interested instead is [the crates.io API](https://crates.io/data-access#api) — in your example, https://crates.io/api/v1/crates/rand probably has the data you're looking for. crates.io implements a superset of the [cargo registry web API](https://doc.rust-lang.org/cargo/reference/registry-web-api.html), so those docs are a good place to start to see what's available.
